charms/jupyter-controller/src/charm.py

```diff
@@ -59,110 +80,181 @@ def __init__(self, *args):
+                    "http": {"url": f"http://localhost:{METRICS_PORT}{METRICS_PATH}"},
                 }
             },
-        )
-        self.model.unit.status = ActiveStatus()
+        }
+
+        return Layer(layer_config)
 
     def _check_leader(self):
+        """Check if this unit is a leader."""
         if not self.unit.is_leader():
-            # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailedError("Waiting for leadership", WaitingStatus)
+            self.logger.info("Not a leader, skipping setup")
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
 
-    def _check_image_details(self):
+    def _check_and_report_k8s_conflict(self, error):
+        """Return True if error status code is 409 (conflict), False otherwise."""
+        if error.status.code == 409:
+            self.logger.warning(f"Encountered a conflict: {error}")
+            return True
+        return False
+
+    def _apply_k8s_resources(self, force_conflicts: bool = False) -> None:
+        """Apply K8S resources.
+
+        Args:
+            force_conflicts (bool): *(optional)* Will "force" apply requests causing conflicting
+                                    fields to change ownership to the field manager used in this
+                                    charm.
+                                    NOTE: This will only be used if initial regular apply() fails.
+        """
+        self.unit.status = MaintenanceStatus("Creating K8S resources")
+        try:
+            self.k8s_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying K8S resources
+                # re-apply K8S resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying K8S resources")
+                self.logger.warning("Apply K8S resources with forced changes against conflicts")
+                self.k8s_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("K8S resources creation failed") from error
+        try:
+            self.crd_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying CRD resources
+                # re-apply CRD resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying CRD resources")
+                self.logger.warning("Apply CRD resources with forced changes against conflicts")
+                self.crd_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("CRD resources creation failed") from error
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
```

Comment:
These two try/excepts could be extracted to a single helper, like (ignoring any typos...):

```python
def _apply_k8s_resources():
  _apply_and_resolve_conflicts(self.k8s_resource_handler, name="K8S")
  _apply_and_resolve_conflicts(self.crd_resource_handler, name="CRD")

def _apply_and_resolve_conflicts(self, resource_handler, name):
        self.unit.status = MaintenanceStatus(f"Creating {name} resources")
        try:
            resource_handler.apply()
        except ApiError as error:
            if self._check_and_report_k8s_conflict(error) and force_conflicts:
                # conflict detected when applying resources
                # re-apply resources with forced conflict resolution
                self.unit.status = MaintenanceStatus(f"Force applying {name} resources")
                self.logger.warning(f"Apply {name} resources with forced changes against conflicts")
                resource_handler.apply(force=force_conflicts)
            else:
                raise GenericCharmRuntimeError(f"{name} resources creation failed") from error
```

I dont know if this helps readability or not.  Might help testability - can at least make some easy unit tests for `_apply_and_resolve_conflicts`

Reply:
This looks great! If there are no objections on this I will re-do it the way you suggested.

Reply:
Will have to leave it for re-design with base class.